socks/socks.go

```diff
@@ -163,29 +164,65 @@ func ParseAddr(s string) Addr {
 
 // Handshake fast-tracks SOCKS initialization to get target address to connect.
 func Handshake(rw io.ReadWriter) (Addr, error) {
-	// Read RFC 1928 section 4 for request and reply structure and sizes
-	buf := make([]byte, MaxReqLen)
+	// Read RFC 1928 for request and reply structure and sizes
 
-	_, err := rw.Read(buf) // SOCKS version and auth methods
-	if err != nil {
+	buf := make([]byte, MaxReqLen)
+	// read VER, NMETHODS, METHODS
+	if _, err := io.ReadFull(rw, buf[:2]); err != nil {
 		return nil, err
 	}
-
-	_, err = rw.Write([]byte{5, 0}) // SOCKS v5, no auth required
-	if err != nil {
+	if buf[0] != 5 { // VER: 5
+		return nil, fmt.Errorf("SOCKS bad version: %d", buf[0])
+	}
+	len := buf[1]
+	if _, err := io.ReadFull(rw, buf[:len]); err != nil {
 		return nil, err
 	}
+	noAuthExist := false
+	for _, method := range buf[:len] {
+		if method == 0 { // METHOD: NO AUTHENTICATION REQUIRED
+			noAuthExist = true
+			break
+		}
+	}
+	if !noAuthExist {
+		return nil, fmt.Errorf("SOCKS no method noAuth: %v", buf)
+	}
 
-	n, err := rw.Read(buf) // SOCKS request: VER, CMD, RSV, Addr
-	if err != nil {
+	// write VER METHOD
+	if _, err := rw.Write([]byte{5, 0}); err != nil {
 		return nil, err
 	}
-	buf = buf[:n]
 
+	// read VER CMD RSV ATYP DST.ADDR DST.PORT
+	if _, err := io.ReadFull(rw, buf[:4]); err != nil {
+		return nil, err
+	}
 	if buf[1] != CmdConnect {
 		return nil, ErrCommandNotSupported
 	}
+	buf = buf[3:]
+	var err error
+	switch buf[0] {
```

Comment:
On the second thought, you were right in the previous commit. It's actually better to call `ReadAddr` here instead of duplicating the same error-prone code path.